react/states_and_effects/introduction_to_state.md

```diff
@@ -0,0 +1,101 @@
+
+Adding more state variables should be as easy as adding more `useState` calls.
+
+Great, you can now use state! But, what's actually going on under the hood? 
+
+### How does State Work in React? 
+
+Let's hit you with some theory.
+
+In React, when a component's state or props change, the component is destroyed and recreated from scratch. Yes, you heard that right: _destroyed_. This includes the variables, functions, and React nodes. The entire component is recreated but this time the latest state value will be returned from `useState`. This process is called rerendering. Rerendering is a key feature of React that enables it to efficiently update the user interface in response to changes in the underlying data.
+
+<div class="lesson-note" markdown="1">
+
+#### React Reconciliation Algorithm
+
+The process of rerendering generates a new virtual DOM (Document Object Model) tree. The virtual DOM is a lightweight representation of the actual DOM that React uses to keep track of the current state of the UI. React then compares the new virtual DOM tree to the previous one and calculates the minimal set of changes needed to update the actual DOM. This is the reconciliation algorithm. 
+
+</div>
+
+#### Explaining rerendering with the example above 
+
+Whenever `setBackgroundColor` is called, our `App` component is rerendered. Essentially, the component is recreated which means the `onButtonClick` function and our `div` and `button`'s are recreated as well. You might wonder, shouldn't the `backgroundColor` state be recreated as well? Well, React takes the responsibility of keeping track of the latest state and providing it to the component. The initial state value is only used for the component's first render, and is ignored on subsequent renders. 
+
+![code illustration of rerendering](./introduction_to_state/imgs/00.png)
+
+What we covered in this section is just a brief overview of a portion of React internals. You may be curious to learn more, but be advised, this is a deep rabbit hole. What we covered above is enough to get you through the React course. 
+
+### Hooks 
+
+Hooks are functions that let you use React features. All hooks are recognizable by the `use` prefix. For example, `useState` is a hook. We'll use more of these as we get further into the course. For now, remember that hooks have rules that we need to abide by:
+
+1. Hooks can only be called from the top level of a functional component.
+1. Hooks can't be called from inside loops or conditions.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Read [State: A Component's Memory from the React docs](https://react.dev/learn/state-a-components-memory).
```

Comment:
I feel like all of the first three of the react docs sections on `Adding interactivity` are discussed in this lesson. So maybe including read throughs of the other two (`Render and commit` and `State as a Snapshot`) may be good decision?